tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/localedistance/DistanceTable.java

```diff
@@ -0,0 +1,556 @@
+         * If the node already exists, it is simply returned, otherwise a new node is created
+         * and any existing wildcard mappings are copied into it.
+         */
+        private Node getIntermediateNode(MappingKey key) {
+            Node node = getNode(key);
+            if (node == null) {
+                // This is expected to succeed because match rules are given in length
+                // order (i.e. language only before language+script etc.) and we always
+                // expect each group to end with an <ANY,ANY> mapping for the default
+                // distance. Thus, for any longer match rule, we should find (at least)
+                // the <ANY,ANY> node when looking for intermediate nodes.
+                //
+                // NOTE: Currently (desired==ANY) if-and-only-if (supported=ANY), so the
+                // only non-exact match we can get here is the <ANY,ANY> node. If we ever
+                // allow a mix of wildcard/non-wildcard keys, replace the getAnyNode() call
+                // with something like the line below:
+                // ----
+                // Node wildcardMatch = Iterables.find(
+                //      asList(getNode(desired, ANY), getNode(ANY, supported), getNode(ANY,ANY)),
+                //      Objects::nonNull);
+                // ----
+                Node wildcardMatch = getAnyNode();
+                checkState(wildcardMatch != null, "missing <ANY,ANY> mapping: %s", this);
+                // Default distances are the distance between any two *different* unknown
+                // subtags (so if the subtags are the same, the distance is zero).
+                int distance = key.getDesired().equals(key.getSupported()) ? 0 : wildcardMatch.distance;
+                node = new Node(distance);
+                node.copySubtablesFrom(wildcardMatch);
+                subtables.put(key.getDesired(), key.getSupported(), node);
+            }
+            return node;
+        }
+
+        /** Copies all subtable mappings from the given node into this one. */
+        private void copySubtablesFrom(Node src) {
+            checkState(subtables.isEmpty());
+            src.subtables.cellSet().forEach(
+                    c -> subtables.put(c.getRowKey(), c.getColumnKey(), new Node(c.getValue().distance)));
+        }
+
+        /**
+         * Writes all the mappings in the distance table sequentially to given Trie in sorted
+         * table order.
+         *
+         * <p>Mappings are written in a top-down recursive visitation with sub-tables inheriting
+         * the current prefix from parent tables via the given Trie span. At each level any
+         * mapped distances are written before recursing into the sub-tables.
+         */
+        private void writeTo(Span trie) {
+            if (distance >= 0 && (distance & DISTANCE_SKIP_SCRIPT) != 0) {
+                // If a node has no distance set, or has been explicitly marked as "skippable",
```

Comment:
For "... has no distance set, or..." should the test not be `distance < 0 ||` rather than `distance >= 0 &&`   ?

Reply:
Yes ... and now I'm wondering why a test didn't catch that (but I'm also worried the comment could be wrong).
Does the comment seem correct to you? I've changed the code and will add a test if that's feasible, but if the comment seems suspicious to you, let me know.

Reply:
Hmm, so all the other change are fine, but if I change the code to match the comment, things break (null pointers).

The code "as is" (using ">= 0" and "&&") passes all tests, but this looks like it can result in a distance of -1 being written in writeDistancePlusSubtables() on about line 369. However I'm honestly not sure if that's okay, or if every distance value in the written Trie must be >= 0 ?

This obviously needs a closer look, and you may well be able to help me untangle the intent here.

Reply:
I've sent you an email about this.